### ISSUE TITLE
Replace PSPs with PSS

### ISSUE THREAD

**Issue by MartaD** [giantswarm/roadmap#256]
PSPs will become deprecated in Kubernetes 1.21 and completely removed in version 1.25. By that time we need to implement another solution to replace PSPs.

```[tasklist]
### Tasks
- [x] test 1.9
- [x] roll 1.9 out to MCs
- [x] https://github.com/giantswarm/giantswarm/issues/22657
- [ ] https://github.com/giantswarm/giantswarm/issues/25146
- [x] write up our suggestion for exception management https://github.com/giantswarm/giantswarm/issues/26357
- [x] teams start adding new exceptions (there is a dashboard in Grafana cloud showing the current compliance state per-team)
- [ ] https://github.com/giantswarm/giantswarm/issues/26857
```

**Comment by MartaD** [comment]
Giant Swarm is closely watching and contributing to the upstream development of a replacement solution. We will update this ticket once the new approach is agreed on and we'll start adopting it.

**Comment by puja108** [comment]
By now with Kubernetes 1.22 upstream has released the Pod Security Admission (PSA) feature in alpha.

We are still not recommending using Pod Security Admission over PSPs, yet. They can both enforce the same Pod Security Standards (sth we already do), but with PSP you can define exemptions easily with RBAC and PSP dynamically, with PSA you have to add exemptions to the configuration of the admission controller statically.

On top of that an exemption in PSA suddenly means you can skip ALL the security rules not just a list of things you want to enable for this specific workload.

TLDR PSA is super early, We recommend waiting another 1-2 more releases before migrating away from PSP. PSP only goes away with 1.25 and that is a year out. Until then we have a better picture how to ensure similar security with either PSA or other tools like Kyverno, so I'm for now keeping the blocked label.

cc @stone-z

**Comment by puja108** [comment]
I think we might need to keep this as epic. We then might need a few issues to track the progress:
1. Have (kyverno) PSS app ready and documented
2. Migrate/adapt all apps to support 1
3. Migrate existing cluster to 1
4. Make 1 the default?

cc @alex-dabija @cornelius-keller for coordination with KaaS/CAPI work

**Comment by puja108** [comment]
@stone-z do we now with Kyverno 1.9 and the exception CRD out (https://kyverno.io/blog/2023/02/01/kyverno-1.9-released/#policy-exceptions) have a solution that we would use for PSS exception management? Is the new version rolled out and could Shield already test if your idea how to manage the exceptions would work? If yes, I'd say the next step would be documenting how teams (both at GS and at customers) need to adapt their apps to include the new exception management so we can prioritize that work soon in the respective internal teams and also work with customers to get them prepared for a future upgrade where PSPs are replaced by Kyverno enforced PSS. In case that process is tedious we might need to think about providing additional tooling (or does upstream Kyverno have sth prepped?) to make that easier.

**Comment by stone-z** [comment]
> do we now with Kyverno 1.9 and the exception CRD out have a solution that we would use for PSS exception management?

Yes, at least we believe so.

> Is the new version rolled out

No. It was only released a few days ago and we haven't fully tested it yet. But it'll be on MCs once we have tested it, and from there is when we'll start the migration push.

> next step would be documenting

Yes, we'll definitely be writing and updating docs

> or does upstream Kyverno have sth prepped?

They don't, and they are very evasive about recommending or prescribing a particular approach. They like to just provide options. We have some tooling in mind (see below).

Our next steps will be:
- test 1.9 (last RC we tried, the exception mechanism wasn't fully implemented. If that is still the case, we will need to adjust plans)
- roll 1.9 out to MCs (pretty much immediately)
- [create known exceptions for core components](https://github.com/giantswarm/giantswarm/issues/22657)
- [solve an issue with audit + enforcement modes](https://github.com/giantswarm/giantswarm/issues/25146)
- write up our suggestion for exception management
- teams start adding new exceptions (there is a dashboard in Grafana cloud showing the current compliance state per-team)
- we start changing policies from `audit` to `enforce` on MCs

For migration support, shield will add the first exceptions so there are references for other teams. Based on the current reports, we don't anticipate a large number of exceptions (most GS workloads are pretty simple from a permissions perspective) but if there are more than can be handled in a reasonable amount of time, we can look at generating exceptions based on the current PSPs in the cluster.

**Comment by puja108** [comment]
Cool, I'll move your list to a tasklist in the header of this issue then. Can you add a rough ETA (as Roadmap project, start and end date) to it so we can include it in CAPI planning?

**Comment by stone-z** [comment]
Sure, we can close this one. Rollouts and adoption are underway, nothing is still waiting for implementation
